pypeit/images/rawimage.py

```diff
@@ -19,399 +19,701 @@
+                          spatial_flexure_shift=False,
+                          flatten=False)
 
     @property
     def bpm(self):
         """
-        Generate and return the bad pixel mask for this image
-        Warning:  BPM masks are for processed (e.g. trimmed, rotated) images only!
+        Generate and return the bad pixel mask for this image.
+
+        .. warning::
+
+            BPMs are for processed (e.g. trimmed, rotated) images only!
 
         Returns:
             `numpy.ndarray`_:  Bad pixel mask with a bad pixel = 1
 
         """
         if self._bpm is None:
-            self._bpm = self.spectrograph.bpm(shape=self.image.shape,
-                                    filename=self.filename,
-                                    det=self.det)
+            # TODO: Was this ever being called?
+#            self._bpm = self.spectrograph.bpm(shape=self.image.shape, filename=self.filename,
+#                                              det=self.det)
+            # TODO: Pass bias if there is one?
+            self._bpm = self.spectrograph.bpm(self.filename, self.det, shape=self.image.shape)
         return self._bpm
 
-    def apply_gain(self, force=False):
+    @property
+    def use_flat(self):
         """
-        Apply the Gain values to self.image
+        Return a flag setting if the flat data should be used in the image
+        processing.
+        """
+        if self.par is None:
+            return False
+        # TODO: At the moment, we can only perform any of the flat-field
+        # corrections if we are applying the pixel-flat correction.
+#        return self.par['use_pixelflat'] or self.par['use_specillum'] or self.par['use_illumflat']
+        return self.par['use_pixelflat']
 
-        Args:
-            force (bool, optional):
+    def apply_gain(self, force=False):
+        """
+        Use the gain to convert images from ADUs to electrons/counts.
 
-        Returns:
-            `numpy.ndarray`:  copy of self.image
+        Conversion applied to :attr:`image, :attr:`var`, and :attr:`rn2img`.
 
+        Args:
+            force (:obj:`bool`, optional):
+                Force the gain to be applied to the image, even if the step log
+                (:attr:`steps`) indicates that it already has been.
         """
         step = inspect.stack()[0][3]
-        # Check if already trimmed
-        if self.steps[step] and (not force):
-            msgs.warn("Gain was already applied. Returning")
-            return self.image.copy()
-
-        gain = np.atleast_1d(self.detector['gain']).tolist()
-        # Apply
-        self.image *= procimg.gain_frame(self.datasec_img, gain)
+        if self.steps[step] and not force:
+            # Already applied
+            msgs.warn('Gain was already applied.')
+            return
+        gain = procimg.gain_frame(self.datasec_img, np.atleast_1d(self.detector['gain']))
+        self.image *= gain
+        if self.var is not None:
```

Comment:
Why would we ever have a variance image that was not already in electrons. At the point at which we convert an image from ADU to electrons, we should be computing and delivering the noise model. For that reason I do not ever see a reason to compute a variance image or an rn2 image in units of ADU.

Reply:
Gain application is now done first, so this has been removed.